File: R/hu.R
## Wrapper function for C code to calculate heat units
## M. Espe
## July 2015

##' Calculate heat units
##'
##' Takes a vector of daily minimum and maximum temperatures,
##'     and returns a vector of daily heat units. Heat units are
##'     calculated by summing over approximated hourly temperatures.
##'
##' @title Heat Units
##' @param tmin vector of daily minimum temperatures
##' @param tmax vector of daily maximum temperatures
##' @param t_base base cardinal temperature for development
##' @param t_opt optimimum temperature for development 
##' @param t_high maximum temperature for development 
##'
##' @return vector of daily heat units
##' @author Matt Espe 
heat_units <- function(tmin, tmax,
                       t_base = 8, t_opt = 30, t_high = 42)
{
    if(!all(tmax > tmin))
       stop("Tmax must be greater than associated Tmin")
    if(any(is.na(c(tmin, tmax))))
       stop("NAs not allowed")
           
    .C('heat_units', n = as.integer(length(tmin)),
       t_min = as.numeric(tmin),
       t_max = as.numeric(tmax),
       ans = as.numeric(rep(0, length(tmin))),
       t_base = as.numeric(t_base),
       t_opt = as.numeric(t_opt),
       t_high = as.numeric(t_high),
       cos_vals = as.numeric(cos(0.2618 * -13:10) / 2))$ans
    
}

File: src/hu.c
/*
Function to calculate heat units according to 
the equations in Bouman et al 2001
 */

void
heat_units(int *n, double *t_min, double *t_max, double *ans,
		   double *t_base, double *t_opt, double *t_high,
		   double *cos_vals)
{
  // Initialize values
  int n_rec = n[0];
  int i, h;
  double d_hu, tmp;
   
 
  // Loop over days
  for(i = 0; i < n_rec; i++) {
	// Reset d_hu each day
	d_hu = 0;
     
	// Loop over hours
	for(h = 0; h < 24; h++) {
	  // Calculate hourly heat units
	  tmp = ((t_max[i] + t_min[i])/2) + ((t_max[i] - t_min[i]) * cos_vals[h]);
	  // If statements for accumulation
	  if(tmp <= t_base[0] || tmp >= t_high[0]){
		d_hu = d_hu + 0;
	  }
       
	  if(tmp > t_base[0] && tmp <= t_opt[0]){
		d_hu = d_hu + (tmp - t_base[0])/24;
	  }
       
	  if(tmp > t_opt[0] && tmp < t_high[0]){
		d_hu = d_hu + ((t_opt[0] - (tmp - t_opt[0]) * (t_opt[0] - t_base[0])/(t_high[0] - t_opt[0]))/24);
	  }
	}
     
	ans[i] = d_hu;
  }
}
